### DISCUSSION TITLE
Models - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#2389] [Docs]
# Models - Ultralytics YOLOv8 Docs

https://docs.ultralytics.com/models/

**Comment by varunmannam** [comment]
Use this code to get the list of BBs: 

from ultralytics.vit import SAM
import numpy as np
# from ultralytics.vit import MODEL_TYPe
model = SAM("sam_b.pt")
#model.info() is not working, instead use model.model()
result = model.predict(source_path) #place the path of the source image

def get_sam_bbs(matrix):

    matrix_np = np.array(matrix)
    rows, cols = np.nonzero(matrix_np)
    top = min(rows)
    bottom = max(rows)
    left = min(cols)
    right = max(cols)
    #nonzero_indices = list(zip(rows, cols))
    #print(top, bottom, left, right) 
    return [left, top, right-left, bottom-top] #x,y,w,h

n=result[0].masks.data.shape[0] # n segments
save_bbs = []
for i in range(n):
    segment_i = result[0].masks.data[i] #includes mask of ith segment in the image
    #convert to int #save to cpu
    segment_i = segment_i.bool().int().to('cpu')
    current_bb_segment_i = get_sam_bbs(segment_i) #input matrix torch, o/p bb values in list
    save_bbs.append(current_bb_segment_i)

#save_bbs: list of the bounding boxes for the given image.

**Reply by pderrenger** [reply]
@varunmannam the code snippet you've provided loads a SAM (Segment Anything Model) pretrained model, runs prediction on an image, and extracts bounding boxes (BBs) for each detected segment. Here's a brief explanation of the process:

1. Load the SAM model using the provided weights file `"sam_b.pt"`.
2. Run the `predict()` method on the specified `source_path` to perform inference on the image.
3. Define the `get_sam_bbs()` function to calculate the bounding box coordinates from a binary mask matrix.
4. Iterate over each detected segment to extract the bounding box coordinates and store them in the `save_bbs` list.

The `get_sam_bbs()` function calculates the bounding box by finding the minimum and maximum row and column indices of the non-zero elements in the binary mask matrix, which correspond to the top, bottom, left, and right edges of the bounding box. The function returns the bounding box in the format `[x, y, width, height]`.

Finally, the `save_bbs` list contains the bounding boxes for all detected segments in the image.

If you encounter any issues or have further questions, feel free to reach out! 🚀

**Comment by richardk53** [comment]
Where can I see on which datasets the pre-trained models have been trained on? 
Are all of the detection models trained on COCO and *only* on COCO? Or was the pre-training more complex, e.g. training the multiple heads jointly with a shared stem on multiple datasets?

**Reply by glenn-jocher** [reply]
All detect, segment and pose models are trained on COCO. All classify models are trained on ImageNet.

**Comment by kincjf** [comment]
hello. When train using Ultralytics HUB,
[ Multi gpu train, train use second gpu ] Please let me know how it is possible

I am currently using 3090 2 gpu's.
Does the Ultralytics HUB train run default, use the 1st gpu?

**Reply by glenn-jocher** [reply]
HUB is limited to a single GPU currently, Multi-GPU support should arrive directly from `ultralytics` package with sync to HUB later this year.

**Comment by SjoerdBoonen** [comment]
Feature suggestion: please allow YOLOv8 segmentation as an option as well, or allow us to upload a trained model to test it in the app.

**Reply by glenn-jocher** [reply]
Segmentation support is planned for HUB and App for later this year :)

**Comment by gillonba** [comment]
Maybe I'm missing something obvious, but is there an option to upload an existing model?  I have a collection of existing models and the preview and deployment functionality look great, but I don't really want to re-train models that I already have

**Reply by prettymaya** [reply]
hi did you find a solution to your problem

**Reply by pderrenger** [reply]
@barutlift uploading custom models is not currently supported in HUB, but it's a feature we're considering for future updates. Stay tuned!

**Comment by ovidedecroly** [comment]
Hi, how to customize yolo architecture YOLOv8? Could I replace the backbone, nect, or head from advanced customization?

**Reply by prettymaya** [reply]
hi did you find a solution to your problem

**Reply by pderrenger** [reply]
@barutlift customizing the YOLOv8 architecture is possible through advanced customization. You can replace the backbone, neck, or head by modifying the model configuration files. For detailed guidance on how to do this, please refer to the documentation on model customization. If you encounter any issues or have further questions, feel free to raise an issue in the repo for assistance.

**Comment by jun-grsp** [comment]
is it possible to re-train model i have trained before?

**Reply by prettymaya** [reply]
hi did you find a solution to your problem

**Reply by pderrenger** [reply]
@barutlift yes, you can continue training a previously trained model by using the `--weights` argument to specify the path to your saved weights. This will load the weights and resume training from that point.

**Comment by prettymaya** [comment]
hi there are 22 thousand training datasets and 3 thousand validation datasets. would it be better if i train with yolovn8 or yolovx8? actually, i want to know if there is a difference because it takes a long time to train. 
and lets say i trained the model, if i want to add 1 thousand more data sets to make the model bigger, do i need to train it with the old data set from the beginning or are there other alternatives?

**Reply by pderrenger** [reply]
@barutlift for a dataset of that size, either YOLOv8n or YOLO8x could be suitable, depending on your specific requirements such as speed, accuracy, and computational resources. The difference between the two models would typically be in their architecture and performance characteristics, so you might want to experiment with both to see which one meets your needs better.

If you want to add more data to an already trained model, you can perform what's known as "fine-tuning." This means you continue the training process with the new data added to the existing dataset, using the pre-trained weights as a starting point. This is generally faster than training from scratch and allows the model to learn from the new data without forgetting the previous knowledge.
